src/domain/session/room/timeline/linkify/regex.js

```diff
@@ -0,0 +1,36 @@
+/*
+The regex is split into component strings;
+meaning that any escapes (\) must be also
+be escaped.
+*/
+const scheme = "(?:https|http|ftp):\\/\\/";
+const host = "[a-zA-Z0-9:.\\[\\]-]";
+
+/*
+A URL containing path (/) or fragment (#) component
+is allowed to end with any character which is not 
+space nor punctuation. The ending character may be 
+non-ASCII.
+*/
+const end = "[^\\s.,?!]";
+const additional = `[\\/#][^\\s]*${end}`;
+
+/*
+Similarly, a URL not containing path or fragment must
+also end with a character that is not space nor punctuation.
+Additionally, the ending character must also be ASCII.
+*/
+const nonASCII = "\\u{80}-\\u{10ffff}";
+const endASCII = `[^\\s${nonASCII}.,?!]`;
+
+/*
+Things to keep in mind:
+1.  URL must not contain non-ascii characters in host but may contain
+    them in path or fragment components.
+    https://matrix.org/<smiley> - valid
+    https://matrix.org<smiley> - invalid
+2. Do not treat punctuation at the end as a part of the URL (.,?!)
+*/
+const urlRegex = `${scheme}${host}+(?:${additional}|${endASCII})`;
```

Comment:
endASCII is basically checking the last character of the hostname, right? It should have the same constraints as `host`, and additionally not accept any punctuation. This is true whether there is a path/fragment or not. So could we instead simplify to this?:
```js
const scheme = "(?:https|http|ftp):\\/\\/";
const noSpaceNorPunctuation = "[^\\s.,?!]";
const hostCharacter = "[a-zA-Z0-9:.\\[\\]-]";
// using non-consuming group here to combine two criteria for the last character
const host = `${hostCharacter}*(?=${hostCharacter})${noSpaceNorPunctuation}`;
// use sub groups so we accept just / or #, but if anything comes after it should not end with punctuation
const pathOrFragment = `(?:[\\/#](?:[^\\s]*${noSpaceNorPunctuation})?)`;
// path or fragment is optional
const urlRegex = `${scheme}${host}${pathOrFragment}?`;
```

This way the last character of the host won't accept any characters not accepted by DNS, like `https://matrix.org<`